store/cachemulti/store.go

```diff
@@ -127,6 +127,36 @@ func (cms Store) Write() {
 	}
 }
 
+// Clone creates a snapshot of each store of the cache-multistore.
+// Each copy is a copy-on-write operation and therefore is very fast.
+func (cms Store) Clone() types.CacheMultiStore {
+	stores := make(map[types.StoreKey]types.CacheKVStore, len(cms.stores))
+	for key, store := range cms.stores {
+		stores[key] = store.Clone()
+	}
+	return Store{
+		db:     cms.db.Clone(),
+		stores: stores,
+
+		keys:         cms.keys,
+		traceWriter:  cms.traceWriter,
+		traceContext: cms.traceContext,
+	}
+}
+
+// Restore restores the cache-multistore cache to a given snapshot.
```

Comment:
This method is defined on a value receiver, which means that it operates on a copy of the value it's called against, rather than the original value itself. The method acts on a copy of the value, and any operations on the fields of the type are made against copies of those fields. In short, a method like this, which intends to modify the state of the Store receiver it operates on, must be defined on a pointer receiver. And mutating operations must also be safe for concurrent use, so must acquire an e.g. mutex, etc.

Reply:
The receiver itself is not mutated, but the internal stores it referenced is.

Reply:
That may be true, but accessing those fields must be synchronized with a mutex.

Reply:
those fields are not mutated after construction, I think that's why the whole `cachemulti.Store` don't need to mutex in the first place.

Reply:
At a minimum, the `db` field is mutated by its `Restore` method, which means it must be synchronized somewhere. Since the dbm.DB type itself doesn't guarantee concurrency safety, synchronization must be provided by the caller.

Reply:
`db` here is also a `cachekv.Store`.

Reply:
That can't be assumed, and even if it could be assumed, that type is not synchronized, anyway.